### PR TITLE
feat(completions): autocomplete for function types

### PR DESCRIPTION
**Pull request by juleswritescode**
Autocompletions for functions! Some rules that are applied:

- if we're in a `select` clause, we'll prefer function items over table items
- if we're in an "invocation" (parentheses after the current token), we'll prefer functions 
- if we're in a `from` clause, we'll slightly prefer table items

I refactored the `CompletionRelevance` again; now we can update the score based on the completion item score.

I also added docs to the SchemaCache's `Function` and changed the query. Some properties were `Option`s, even though they should always be present: `id`, `schema` etc. – *please verify if I made any wrong assumptions 🙌*

Some extra stuff:
- added a few additional test helpers (only compiled when we're running tests)
- changed the `CompletionItem.preselected` from `Option<bool>` to a simple `bool` 
- added a `ClauseType` enum

Note: I had to comment out an alphanumeric table test. The query generates a tree-sitter tree with an `ERROR`, and the current node has an `end_byte()` that is at position 17, while the query only has a length of 14... will check in a separate PR, not sure why it even worked before 🤷

### CODE REVIEW

File: crates/pg_completions/src/providers/functions.rs [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+    builder::CompletionBuilder, context::CompletionContext, relevance::CompletionRelevanceData,
+    CompletionItem, CompletionItemKind,
+};
+
+pub fn complete_functions(ctx: &CompletionContext, builder: &mut CompletionBuilder) {
+    let available_functions = &ctx.schema_cache.functions;
+
+    let completion_items: Vec<CompletionItem> = available_functions
+        .iter()
+        .map(|foo| CompletionItem {
```
Comment (psteinroe) [inline crates/pg_completions/src/providers/functions.rs:11]: Nit: can't we loop over available functions and add the item dirextly?

**Review by psteinroe — APPROVED** [review]
lgtm! Two nits, but we can merge 🙏🏼

File: crates/pg_completions/src/context.rs [hunk 2]
```diff
@@ -2,6 +2,34 @@ use pg_schema_cache::SchemaCache;
...
+
+impl From<&str> for ClauseType {
+    fn from(value: &str) -> Self {
+        match value {
+            "select" => Self::Select,
+            "where" => Self::Where,
+            "from" => Self::From,
+            "update" => Self::Update,
+            "delete" => Self::Delete,
+            _ => panic!("Unimplemented ClauseType: {}", value),
```
[the remark this answers is not in the document]
Comment (juleswritescode) [reply crates/pg_completions/src/context.rs:22]: Great point! I liked the throw for tests, but it's much better not to fail in prod

[the remark this answers is not in the document]
Comment (juleswritescode) [reply crates/pg_completions/src/context.rs]: Adjusted, but it still throws in tests – let me know if you'd like to change that

File: crates/pg_completions/src/providers/functions.rs (hunk 1 above)
Comment (juleswritescode) [reply crates/pg_completions/src/providers/functions.rs]: great point, saves one full iteration – adjusted

File: crates/pg_completions/src/context.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (psteinroe) [reply crates/pg_completions/src/context.rs]: That's fine 👍🏼

**Comment by psteinroe** [comment]
@juleswritescode, sorry, can you fix the conflict again? Should have merge the other pr after this one

**Comment by juleswritescode** [comment]
> @juleswritescode, sorry, can you fix the conflict again? Should have merge the other pr after this one

Check!!

**Pull request merged**